rust/rustParser.g4

```diff
@@ -0,0 +1,714 @@
+
+bareFunctionType
+   : forLifetimes? functionQualifiers 'fn' '(' functionParametersMaybeNamedVariadic? ')' bareFunctionReturnType?
+   ;
+bareFunctionReturnType: '->' typeNoBounds;
+functionParametersMaybeNamedVariadic
+   : maybeNamedFunctionParameters
+   | maybeNamedFunctionParametersVariadic
+   ;
+maybeNamedFunctionParameters: maybeNamedParam (',' maybeNamedParam)* ','?;
+maybeNamedParam: outerAttribute* ((identifier | '_') ':')? type;
+maybeNamedFunctionParametersVariadic
+   : (maybeNamedParam ',')* maybeNamedParam ',' outerAttribute* '...'
+   ;
+traitObjectType: 'dyn'? typeParamBounds;
+traitObjectTypeOneBound: 'dyn'? traitBound;
+implTraitType: 'impl' typeParamBounds;
+implTraitTypeOneBound: 'impl' traitBound;
+inferredType: '_';
+
+typeParamBounds: typeParamBound ('+' typeParamBound)* '+'?;
+typeParamBound: lifetime | traitBound;
+traitBound: '?'? forLifetimes? typePath | '(' '?'? forLifetimes? typePath ')';
+lifetimeBounds: (lifetime '+')* lifetime?;
+lifetime: LIFETIME_OR_LABEL | '\'static' | '\'_';
+
+identifier: NON_KEYWORD_IDENTIFIER | RAW_IDENTIFIER;
+keyword
+   : KW_AS
+   | KW_BREAK
+   | KW_CONST
+   | KW_CONTINUE
+   | KW_CRATE
+   | KW_ELSE
+   | KW_ENUM
+   | KW_EXTERN
+   | KW_FALSE
+   | KW_FN
+   | KW_FOR
+   | KW_IF
+   | KW_IMPL
+   | KW_IN
+   | KW_LET
+   | KW_LOOP
+   | KW_MATCH
+   | KW_MOD
+   | KW_MOVE
+   | KW_MUT
+   | KW_PUB
+   | KW_REF
+   | KW_RETURN
+   | KW_SELFVALUE
+   | KW_SELFTYPE
+   | KW_STATIC
+   | KW_STRUCT
+   | KW_SUPER
+   | KW_TRAIT
+   | KW_TRUE
+   | KW_TYPE
+   | KW_UNSAFE
+   | KW_USE
+   | KW_WHERE
+   | KW_WHILE
+
+   // 2018+
+   | KW_ASYNC
+   | KW_AWAIT
+   | KW_DYN
+   // reserved
+   | KW_ABSTRACT
+   | KW_BECOME
+   | KW_BOX
+   | KW_DO
+   | KW_FINAL
+   | KW_MACRO
+   | KW_OVERRIDE
+   | KW_PRIV
+   | KW_TYPEOF
+   | KW_UNSIZED
+   | KW_VIRTUAL
+   | KW_YIELD
+   | KW_TRY
+   | KW_UNION
+   | KW_STATICLIFETIME
+   ;
+macroIdentifierLikeToken
+   : keyword
+   | identifier
+   | KW_MACRORULES
+   | KW_UNDERLINELIFETIME
+   | KW_DOLLARCRATE
+   | LIFETIME_OR_LABEL
+   ;
+macroLiteralToken: literalExpression;
+// macroDelimiterToken: '{' | '}' | '[' | ']' | '(' | ')';
+macroPunctuationToken
+   : '-'
+   //| '+' | '*'
+   | '/'
+   | '%'
+   | '^'
+   | '!'
+   | '&'
+   | '|'
+   | '&&'
+   | '||'
+   | shl
+   | shr
+   | '+='
+   | '-='
+   | '*='
+   | '/='
+   | '%='
+   | '^='
+   | '&='
+   | '|='
+   | '<<='
+   | '>>='
+   | '='
+   | '=='
+   | '!='
+   | '>'
+   | '<'
+   | '>='
+   | '<='
+   | '@'
+   | '_'
+   | '.'
+   | '..'
+   | '...'
+   | '..='
+   | ','
+   | ';'
+   | ':'
+   | '::'
+   | '->'
+   | '=>'
+   | '#'
+   //| '$' | '?'
+   ;
+
+// LA can be removed, legal rust code still pass but the cost is `let c = a < < b` will pass... i hope antlr5 can add
+// some new syntax? dsl? for these stuff so i needn't write it in (at least) 5 language
+
+shl: '<' {_input.LA(1)=='<'}? '<';
+shr: '>' {_input.LA(1)=='>'}? '>';
```

Comment:
Maybe extract these predicates into base class as we made for other grammars (JavaScript, PHP)?

Reply:
I'll do it after test and other stuff ready.